src/librustc_borrowck/borrowck/check_loans.rs

```diff
@@ -745,6 +745,26 @@ impl<'a, 'tcx> CheckLoanCtxt<'a, 'tcx> {
                                                      use_kind, lp_base);
             }
             LpExtend(ref lp_base, _, LpInterior(InteriorField(_))) => {
+                match lp_base.to_type().sty {
+                    ty::ty_struct(def_id, _) | ty::ty_enum(def_id, _) => {
+                        if ty::has_dtor(self.tcx(), def_id) {
+                            // In the case where the owner implements drop, then
+                            // the path must be initialized to prevent a case of
+                            // partial reinitialization
```

Comment:
This code is an improvement on the prior PR that you rebased, thanks!

However, I do wonder whether we could simplify further and just call `self.check_if_path_is_moved(id, span, use_kind, lp_base);` after doing the `ty::has_dtor` check, the same way that the other cases below do.

Does that produce an error message that is too obscure?  (Or, is there some technical reason we cannot do that?)